crates/hc_seed_bundle/src/seed_cipher/seed_bundle.rs

```diff
@@ -0,0 +1,239 @@
+    where
+        S: serde::Serializer,
+    {
+        ISer("hcsb0", &self.cipher_list, &self.app_data).serialize(serializer)
+    }
+}
+
+/// A helper struct for deserializing a seed bundle from the compact msgpack fmt
+#[derive(serde::Deserialize)]
+struct IDes<'lt>(
+    &'lt str,
+    Box<[SeedCipher]>,
+    #[serde(with = "serde_bytes")] Box<[u8]>,
+);
+
+impl<'de> serde::Deserialize<'de> for SeedBundle {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let dec: IDes<'de> = serde::Deserialize::deserialize(deserializer)?;
+        if dec.0 != "hcsb0" {
+            return Err(serde::de::Error::custom(format!(
+                "unsupported bundle version: {}",
+                dec.0
+            )));
+        }
+        Ok(SeedBundle {
+            cipher_list: dec.1,
+            app_data: dec.2,
+        })
+    }
+}
+
+/// The more ergonomic rust structure of an (encrypted) seed cipher.
+#[derive(Debug)]
+pub(crate) enum SeedCipher {
+    /// PwHash type seed cipher
+    PwHash {
+        /// argon salt
+        salt: U8Array<16>,
+
+        /// argon mem limit
+        mem_limit: u32,
+
+        /// argon ops limit
+        ops_limit: u32,
+
+        /// secretstream header
+        header: U8Array<24>,
+
+        /// secretstream cipher
+        cipher: U8Array<49>,
+    },
+    /// Security Questions type seed cipher
+    SecurityQuestions {
+        /// argon salt
+        salt: U8Array<16>,
+
+        /// argon mem limit
+        mem_limit: u32,
+
+        /// argon ops limit
+        ops_limit: u32,
+
+        /// the three security questions to ask the user
+        question_list: (String, String, String),
+
+        /// secretstream header
+        header: U8Array<24>,
+
+        /// secretstream cipher
+        cipher: U8Array<49>,
+    },
+}
+
+impl serde::Serialize for SeedCipher {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        // serialize into the more compact msgpack array format
+        match self {
+            Self::PwHash {
+                salt,
+                mem_limit,
+                ops_limit,
+                header,
+                cipher,
+            } => ("pw", salt, mem_limit, ops_limit, header, cipher)
+                .serialize(serializer),
+            Self::SecurityQuestions {
+                salt,
+                mem_limit,
+                ops_limit,
+                question_list,
+                header,
+                cipher,
+            } => (
+                "qa",
+                salt,
+                mem_limit,
+                ops_limit,
+                &question_list.0,
+                &question_list.1,
+                &question_list.2,
+                header,
+                cipher,
+            )
+                .serialize(serializer),
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SeedCipher {
```

Comment:
Can you add some roundtrip tests?

Reply:
The tests are all in the final PR https://github.com/holochain/lair/pull/49